[REG 2.067.0] Issue 14431 - huge slowdown of compilation speed

9rnsr opened the pull request:
This is same with #4784, excepting the base branch is 'stable'.

---

https://issues.dlang.org/show_bug.cgi?id=14431

In the pull request #4384, all instance has been changed to invoke
semantic3(). It was for the link-failure issue in specific case, but it
was too excessive.
1. Semantic analysis strategy for template instances:
   We cannot determine which instance does not need to be placed in object
   file until semantic analysis completed. Therefore, for all templates
   instantiated in root module, compiler should invoke their semantic3 --
   regardless of whether those are also instantiated in non-root module. If
   a template is _only_ instantiated in non-root module, we can elide its
   semantic3 (and for the compilation speed we should do that).
2. Code generation storategy for template instances:
   If a template is instantiated in non-root module, compiler usually does
   not have to put it in object file. But if a template is instantiated in
   both of root and non-root modules which mutually import each other, it
   needs to placed in objfile.

---

By fixing excessive semantic3 for template instances, we can also fix few regressions which come from the same root.
- [Issue 14508](https://issues.dlang.org/show_bug.cgi?id=14508) - [REG2.067.0] compiling with -unittest instantiates templates in non-root modules
- [Issue 14564](https://issues.dlang.org/show_bug.cgi?id=14564) - [REG2.067] dmd -property -unittest combination causes compiler error
- [Issue 14901](https://issues.dlang.org/show_bug.cgi?id=14901) - [reg 2.067/2.068] template static shared this() run multiple times with separate compilation

9rnsr committed:
fix Issue 14431 - huge slowdown of compilation speed

In the pull request #4384, all instance has been changed to invoke
semantic3(). It was for the link-failure issue in specific case, but it
was too excessive.

1. Semantic analysis strategy for template instances:
  We cannot determine which instance does not need to be placed in object
file until semantic analysis completed. Therefore, for all templates
instantiated in root module, compiler should invoke their semantic3 --
regardless of whether those are also instantiated in non-root module. If
a template is _only_ instantiated in non-root module, we can elide its
semantic3 (and for the compilation speed we should do that).

2. Code generation strategy for template instances:
  If a template is instantiated in non-root module, compiler usually does
not have to put it in object file. But if a template is instantiated in
both of root and non-root modules which mutually import each other, it
needs to placed in objfile.

rainers committed:
Merge pull request #4814 from 9rnsr/fix14541

Issue 14541 - "duplicate COMDAT" linker error with the template forward reference in Tuple.opAssign

9rnsr committed:
fix Issue 14901 - template static shared this() run multiple times with separate compilation

MartinNowak commented on src/struct.c:
```diff
@@ -100,8 +100,12 @@ void semanticTypeInfo(Scope *sc, Type *t)
+            if (TemplateInstance *ti = sd->isInstantiated())
+            {
+                if (ti->minst && !ti->minst->isRoot())
+                    Module::addDeferredSemantic3(sd);
```
Makes sense b/c non-root instances no longer run semantic3.
In the longer run the TypeInfo should only be generated where the struct is instantiated.

MartinNowak commented:
Unfortunately this causes a linker error with [Higgs](https://github.com/higgsjs/Higgs)' test.
```
higgs-test.o:ffi/tests.d:function _D3std5range14__T6strideTAkZ6strideFAkmZ6Result7popBackMFNaNbNiNfZv: error: undefined reference to '_D3std5range10primitives16__T8popBackNTAkZ8popBackNFNaNbNiNfKAkmZm'
higgs-test.o:ffi/tests.d:function _D3std5range14__T6strideTAkZ6strideFAkmZ6Result6moveAtMFNaNbNiNfmZk: error: undefined reference to '_D3std5range10primitives16__T6moveAtTAkTmZ6moveAtFNaNbNiNfAkmZk'
```

I'll try to provide more details, but it seems that some phobos unittest (std.uni) gets compiled into the binary but some templates instantiated by it are not. Might be the `-debug`, `-unittest` issues mentioned in needsCodegen.

MartinNowak commented:
Here is code from Higgs that instantiates the right templates to trigger the issue.
```d
module bug;
void foo()
{
    import std.regex;
    enum hexRegex = ctRegex!("");
}
void main() {}
```
`dmd -debug bug`

The problem itself seems to come from [dbgVerifySorted](https://github.com/D-Programming-Language/phobos/blob/8bf9d9fce4e1918bb4e24ef8eec6fee63d5238f9/std/range/package.d#L7204).

9rnsr commented:
@MartinNowak If a source file is conditionally compiled from two places by using different conditions (`debug` and `version`, including `version(unittest)`), the generate object files are essentially incompatible and link may fail. Currently it's unavoidable, and this PR is completely unrelated to that issue.

OT: In #4626, I had tried to reduce such the link-failure problem for instantiated code...

MartinNowak commented:
From a practical standpoint we cannot do something that breaks programs using ctRegex though.

The problem seems to be a bit more complex though. dbgVerifySorted actually does instantiate stride, it's stride.Result.moveAt instantiating std.range.moveAt which is a problem here. For some reason no code is generated for the latter.

9rnsr commented:
It's not so complex. `dbgVerifySorted` function will have different attributes depending on `-debug` switch. By that, its direct/indirect callers will also have different attributes. If one of the callers is already instantiated in the library (e.g. `SortedRange`), compiler may elide its codegen to use the code in lib file, then it may cause link failure.

When we try to speed up compilation by avoiding more codegen for the instances, it's equivalent that we try to use more many pre-compiled code in the library file. In there, an implicit assumption exists: "the library and the currently compiled code are analyzed under the same conditional compilation".
The excessive semantic analysis and codegen since 2.067 would have been hide the issue - because it's equivalent that more many code will be compiled under the "current condition".

So, this is unavoidable. By taking the compilation speed, we will have to strictly follow the rule of conditional compilation.

MartinNowak commented:
> When we try to speed up compilation by avoiding more codegen for the instances, it's equivalent that we try to use more many pre-compiled code in the library file. In there, an implicit assumption exists: "the library and the currently compiled code are analyzed under the same conditional compilation".

That is understood, but not exactly the problem we're seeing here.

![template_instance](https://cloud.githubusercontent.com/assets/288976/9452974/ac19b0b4-4ab7-11e5-8dde-9c345c6c53b5.png)

Both stride and moveAt are only present b/c the non-root module was analyzed with `-debug`.
If we had a top to bottom instantiation graph (we already have a bottom to top one with `tinst`) we could "taint" all the templates reachable from roots.
It also seems that all the children of CodepointSet should be treated as speculative b/c we don't generate code for CodepointSet and therefor don't know it's real children.

Taking into account the possibly different attributes inferred we apparently have 2 solutions. Generating all instances by default or only linking code that was compiled with the same arguments. The latter required debug/unittest/release phobos libraries or careful avoidance of differences.

9rnsr commented:
> It also seems that the children of CodepointSet should be treated as speculative b/c we don't generate code for CodepointSet and therefor don't know it's real children.

CTFE will invoke semantic3 pass for most of the children of CodepointSet. Therefore currently, compiler can know that SortedRange is its real children.

In other words, for the compilation speed, compiler maximally uses the assumption of the conditional compilation (== pre-compiled library and user code are compiled under the same condition).
It comes from the original design in #2550.

If we adjust the balance of compile speed and link-ability, it should go into future enhancement.

MartinNowak commented:
Interestingly enough, we only generate code for stride b/c of this "hack" https://github.com/D-Programming-Language/dmd/blob/97449e602b346f2dd46a06f74cec74ce3e82d173/src/dtemplate.d#L6329.

9rnsr commented:
@MartinNowak Ah, I finally got the issue you've mentioned. Today, when `-debug` or `-unittest` is specified, the instantiation strategy is switched to the old thing for the `-allinst`. However, as we already know, it's problematic.
That we should do is: if -allinst is not actually specified, we should always use the new strategy even if -debug or -unittest is specified; distinguish root and non-root instances, and have to change some non-root instances to the root ones.

9rnsr commented:
Please give me some days. I'll start to work for that.

9rnsr committed:
Prefer instantiations from root modules when -debug/-unittest is specified

9rnsr committed:
Elide codegen if TypeInfo is speculative

Even if a struct is defined as non-root symbol, some built-in operations
request its TypeInfo. For those, today TypeInfo_Struct is generated in COMDAT.

On the other hand, TypeInfoDeclaration is always generated when the TypeInfo is requested,
even if the code exists in speculative scope. By that, some unneeded
TypeInfo objects had excessively generated.
If a TypeInfo is not actually used, we should stop its generation.

To satisfy two requirements, I added "speculative TypeInfo" concept
associated with template instances.

By this change, when a TypeInfo is *requested*, the codegen pass still
visit it always. But if the corresponding type is instantiated struct or
its derived, and is definitely speculative, the TypeInfo code also won't
be emit into object file.

9rnsr committed:
More TypeInfo objects would be requested from inlined functions

1. When a non-root function is succeeded to expand for inlining,
the additional TypeInfo requests should be handled properly.

2. When a function's semantic3 pass will get called from `canInline`, it
would generate more speculative instances when -debug/-unittest is specified.
If the additionally instantiated function succeeds to expand, it should be
changed to non-speculative, because its code is reachable from the final executable.

9rnsr committed:
(will be squashed later) Better appendToModuleMember implementation and update comments

9rnsr commented:
OK, ready to review.

- [(will be squashed later) ...](7cced26789de95e2e14c1452c7334d50f68f9deb)

   More better implementation. Later it will be squashed into 6c14fadca403a61a578795ec0fb71f9340c7b543.

- [Prefer instantiations from root modules ...](a587c3d6a49c198822a0629071c401e62d592787)

    The core part of the idea. By default, non-root instantiations are preferred to reduce generated code, but when -debug/-unittest is specified, the root instances are preferred to maximize link-ability.

- [Elide codegen if TypeInfo is speculative](https://github.com/9rnsr/dmd/commit/cf56e5ad84164276f31f1d567ded97738a42191d) & [More TypeInfo objects would be requested ...](https://github.com/9rnsr/dmd/commit/0b45d187c8ddb7442e4f7a9fc8780bb1ac58dbbc)

    Fix for the coming out link-failure issues around TypeInfo generation. Even if a TypeInfo is *requested*, it's codegen might be elided.

- [Merge pull request #4814 from 9rnsr/fix14541](cd29d5f34cbacebd83ecdb9727ab1b93f7fa6af1)

    Fix for the coming out link-failure issue in [Win64 phobos unittest](https://auto-tester.puremagic.com/show-run.ghtml?projectid=14&runid=1712974&isPull=true). I cherry-picked the commit from master branch, so the bug is already fixed on that.

9rnsr committed:
Fix isSpeculativeType logic

MartinNowak commented:
I tested this with a bunch of the most downloaded packages and couldn't find any issue.
It does improve build times a bit for all packages and a lot for projects using single file compilation.
The solution makes the template instantiation slightly more complex, which is unfortunate, but at least we can fix recent regressions.

MartinNowak commented:
Auto-merge toggled on

MartinNowak commented:
Unfortunately I hit a linking issue when building [chmgen](https://github.com/D-Programming-Language/dlang.org/blob/22765cd700805f7c331e1464a9635d25ba3c469c/chmgen.d).
Here is a "reduced" test-case.

```d
import std.file;

struct Only(T)
{
    private T _val;
}

auto only(V)(V v)
{
    return Only!V(v);
}

static struct Chain(R...)
{
    R source;
}

auto chain(R...)(R rs)
{
    return Chain!R(rs);
}

void main()
{
    string docRoot;

    chain(
        dirEntries(docRoot , "*.*"   , SpanMode.shallow),
        only(docRoot)
    );
}
```

Seems like TypeInfo for `const(FilterResult!(DirIterator)))` is missing. The TypeInfo for the non-const type is in libphobos2.a.

9rnsr commented:
@MartinNowak 

[Issue 14985](https://issues.dlang.org/show_bug.cgi?id=14985) - [REG2.068.1-b1] Link failure for const TypeInfo of speculative instantiated struct
https://github.com/D-Programming-Language/dmd/pull/4995

MartinNowak commented:
Seems like we have another regression.
[Issue 15021 – [Reg 2.068.1] linker error with speculative instantiation and -inline](https://issues.dlang.org/show_bug.cgi?id=15021)

MartinNowak commented:
And another one [Issue 15030 – [Reg 2.068.1] ICE with recursive delegate, -unittest, and std.range](https://issues.dlang.org/show_bug.cgi?id=15030).

MartinNowak commented:
This one also [Issue 15032 – [REG2.068.1] coverage output is discreted around the calls to map(), canFind(), filter()](https://issues.dlang.org/show_bug.cgi?id=15032).

CyberShadow commented:
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=15985

CyberShadow commented:
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=17398
